core/arch/arm/plat-imx/pm/cpuidle-imx7d.c

```diff
@@ -0,0 +1,238 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2018 NXP
+ *
+ * Peng Fan <peng.fan@nxp.com>
+ */
+
+#include <arm.h>
+#include <arm32.h>
+#include <console.h>
+#include <drivers/imx_uart.h>
+#include <io.h>
+#include <imx.h>
+#include <imx_pm.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/generic_boot.h>
+#include <kernel/interrupt.h>
+#include <kernel/misc.h>
+#include <kernel/panic.h>
+#include <kernel/pm_stubs.h>
+#include <kernel/spinlock.h>
+#include <mm/core_mmu.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <stdint.h>
+#include <sm/optee_smc.h>
+#include <sm/psci.h>
+#include <sm/pm.h>
+#include <sm/sm.h>
+#include <string.h>
+#include <tee/entry_std.h>
+#include <tee/entry_fast.h>
+#include <util.h>
+
+int imx7d_cpuidle_init(void)
+{
+	uint32_t lpm_idle_ocram_base =
+		core_mmu_get_va(TRUSTZONE_OCRAM_START +
+				LOWPOWER_IDLE_OCRAM_OFFSET,
+				MEM_AREA_TEE_COHERENT);
+	struct imx7_pm_info *p =
+		(struct imx7_pm_info *)lpm_idle_ocram_base;
+
+	pm_imx7_iram_tbl_init();
+
+	dcache_op_level1(DCACHE_OP_CLEAN_INV);
+
+	p->va_base = lpm_idle_ocram_base;
+	p->pa_base = TRUSTZONE_OCRAM_START + LOWPOWER_IDLE_OCRAM_OFFSET;
+	p->tee_resume = (paddr_t)virt_to_phys((void *)(vaddr_t)v7_cpu_resume);
+	p->pm_info_size = sizeof(*p);
+	p->ddrc_va_base = core_mmu_get_va(DDRC_BASE, MEM_AREA_IO_SEC);
+	p->ddrc_pa_base = DDRC_BASE;
+	p->ccm_va_base = core_mmu_get_va(CCM_BASE, MEM_AREA_IO_SEC);
+	p->ccm_pa_base = CCM_BASE;
+	p->anatop_va_base = core_mmu_get_va(ANATOP_BASE, MEM_AREA_IO_SEC);
+	p->anatop_pa_base = ANATOP_BASE;
+	p->src_va_base = core_mmu_get_va(SRC_BASE, MEM_AREA_IO_SEC);
+	p->src_pa_base = SRC_BASE;
+	p->iomuxc_gpr_va_base = core_mmu_get_va(IOMUXC_GPR_BASE,
+						MEM_AREA_IO_SEC);
+	p->iomuxc_gpr_pa_base = IOMUXC_GPR_BASE;
+	p->gpc_va_base = core_mmu_get_va(GPC_BASE, MEM_AREA_IO_SEC);
+	p->gpc_pa_base = GPC_BASE;
+	p->gic_va_base = core_mmu_get_va(GIC_BASE, MEM_AREA_IO_SEC);
+	p->gic_pa_base = GIC_BASE;
+
+	p->num_lpi_cpus = 0;
+	p->num_online_cpus = -1;
+
+	memcpy((void *)(lpm_idle_ocram_base + sizeof(*p)),
+	       (void *)(vaddr_t)imx7d_low_power_idle,
+	       LOWPOWER_IDLE_OCRAM_SIZE - sizeof(*p));
+
+	dcache_clean_range((void *)lpm_idle_ocram_base,
+			   LOWPOWER_IDLE_OCRAM_SIZE);
+	/*
+	 * Note that IRAM IOSEC map, if changed to MEM map,
+	 * need to flush cache
+	 */
+	icache_inv_all();
+
+	return 0;
+}
+
+static int lowpoweridle_init;
+
+static void atomic_set(volatile uint32_t *p, uint32_t v)
```

Comment:
Re. the checkpatch warning on this line: I don't think `volatile` is actually useful here, because of the `dmb()` memory barrier. Same in `atomic_read()` below.

Reply:
Why not use `atomic_store_u32()` from `<atomic.h>` instead?

Reply:
@jenswi-linaro +1
I was about to ask, then I considered that the memory order specified in our `atomic_store_u32()` (`__ATOMIC_RELAXED`) is weaker than the one expressed in this code (`dmb()`). But given the purpose, "relaxed" may be good too. Let's see what @MrVan thinks.

Reply:
I was not aware of the api. it is ok to use atomic_store_u32 here :) Thanks.